src/dal/views.py

```diff
@@ -92,11 +103,55 @@ def get_queryset(self):
         """Filter the queryset with GET['q']."""
         qs = super(BaseQuerySetView, self).get_queryset()
 
-        if self.q:
-            qs = qs.filter(**{'%s__icontains' % self.model_field_name: self.q})
+        qs = self.get_search_results(qs, self.q)
 
         return qs
 
+    def get_search_fields(self):
+        """Get the fields to search over."""
+        if self.search_fields:
+            return self.search_fields
+        else:
+            return [self.model_field_name]
+
+    def _construct_search(self, field_name):
+        """Apply keyword searches."""
+        if field_name.startswith("^"):
+            return "%s__istartswith" % field_name[1:]
+        elif field_name.startswith("="):
+            return "%s__iexact" % field_name[1:]
+        elif field_name.startswith("@"):
+            return "%s__search" % field_name[1:]
+        else:
+            return "%s__icontains" % field_name
```

Comment:
Do we really need a code with `^`, `=` and `@` ?
Otherwise, we could just let users set `search_fields = foo__icontains` and so on.

Reply:
Just checking in if you would like me to make this change so this can get merged in

> I do make use of them in my application and they are supported in the
Django admin.  If you prefer they be kept out of the base library, I will
amend the PR.

Reply:
I suppose both work in this case so it'd be fine except that there is no test in this patch so if you depend on it then keep in mind another contributor might break it in the future.

Reply:
Looking at the existing tests, they all seem to be selenium based?  I am not familiar with these types of tests - is their any set up for writing some simple unit tests for the changes I made?

Reply:
Actually this code comes from django and has also had some updates:

```0bcdcc7eb9c (Loic Bistuer             2013-08-03 23:15:15 +0700  986)
470a9bb22d4 (Loic Bistuer             2013-08-04 17:18:17 +0700  987)     def get_search_results(self, request, queryset, search_term):
470a9bb22d4 (Loic Bistuer             2013-08-04 17:18:17 +0700  988)         """
5411821e3b8 (Anton Samarchyan         2017-01-24 15:31:57 -0500  989)         Return a tuple containing a queryset to implement the search
470a9bb22d4 (Loic Bistuer             2013-08-04 17:18:17 +0700  990)         and a boolean indicating if the results may contain duplicates.
470a9bb22d4 (Loic Bistuer             2013-08-04 17:18:17 +0700  991)         """
470a9bb22d4 (Loic Bistuer             2013-08-04 17:18:17 +0700  992)         # Apply keyword searches.
470a9bb22d4 (Loic Bistuer             2013-08-04 17:18:17 +0700  993)         def construct_search(field_name):
470a9bb22d4 (Loic Bistuer             2013-08-04 17:18:17 +0700  994)             if field_name.startswith('^'):
470a9bb22d4 (Loic Bistuer             2013-08-04 17:18:17 +0700  995)                 return "%s__istartswith" % field_name[1:]
470a9bb22d4 (Loic Bistuer             2013-08-04 17:18:17 +0700  996)             elif field_name.startswith('='):
470a9bb22d4 (Loic Bistuer             2013-08-04 17:18:17 +0700  997)                 return "%s__iexact" % field_name[1:]
470a9bb22d4 (Loic Bistuer             2013-08-04 17:18:17 +0700  998)             elif field_name.startswith('@'):
470a9bb22d4 (Loic Bistuer             2013-08-04 17:18:17 +0700  999)                 return "%s__search" % field_name[1:]
244cc401559 (Krzysztof Nazarewski     2017-07-05 13:00:10 +0200 1000)             # Use field_name if it includes a lookup.
244cc401559 (Krzysztof Nazarewski     2017-07-05 13:00:10 +0200 1001)             opts = queryset.model._meta
244cc401559 (Krzysztof Nazarewski     2017-07-05 13:00:10 +0200 1002)             lookup_fields = field_name.split(LOOKUP_SEP)
244cc401559 (Krzysztof Nazarewski     2017-07-05 13:00:10 +0200 1003)             # Go through the fields, following all relations.
244cc401559 (Krzysztof Nazarewski     2017-07-05 13:00:10 +0200 1004)             prev_field = None
244cc401559 (Krzysztof Nazarewski     2017-07-05 13:00:10 +0200 1005)             for path_part in lookup_fields:
244cc401559 (Krzysztof Nazarewski     2017-07-05 13:00:10 +0200 1006)                 if path_part == 'pk':
244cc401559 (Krzysztof Nazarewski     2017-07-05 13:00:10 +0200 1007)                     path_part = opts.pk.name
244cc401559 (Krzysztof Nazarewski     2017-07-05 13:00:10 +0200 1008)                 try:
244cc401559 (Krzysztof Nazarewski     2017-07-05 13:00:10 +0200 1009)                     field = opts.get_field(path_part)
244cc401559 (Krzysztof Nazarewski     2017-07-05 13:00:10 +0200 1010)                 except FieldDoesNotExist:
244cc401559 (Krzysztof Nazarewski     2017-07-05 13:00:10 +0200 1011)                     # Use valid query lookups.
244cc401559 (Krzysztof Nazarewski     2017-07-05 13:00:10 +0200 1012)                     if prev_field and prev_field.get_lookup(path_part):
244cc401559 (Krzysztof Nazarewski     2017-07-05 13:00:10 +0200 1013)                         return field_name
244cc401559 (Krzysztof Nazarewski     2017-07-05 13:00:10 +0200 1014)                 else:
244cc401559 (Krzysztof Nazarewski     2017-07-05 13:00:10 +0200 1015)                     prev_field = field
244cc401559 (Krzysztof Nazarewski     2017-07-05 13:00:10 +0200 1016)                     if hasattr(field, 'get_path_info'):
244cc401559 (Krzysztof Nazarewski     2017-07-05 13:00:10 +0200 1017)                         # Update opts to follow the relation.
244cc401559 (Krzysztof Nazarewski     2017-07-05 13:00:10 +0200 1018)                         opts = field.get_path_info()[-1].to_opts
244cc401559 (Krzysztof Nazarewski     2017-07-05 13:00:10 +0200 1019)             # Otherwise, use the field with icontains.
244cc401559 (Krzysztof Nazarewski     2017-07-05 13:00:10 +0200 1020)             return "%s__icontains" % field_name
470a9bb22d4 (Loic Bistuer             2013-08-04 17:18:17 +0700 1021)
470a9bb22d4 (Loic Bistuer             2013-08-04 17:18:17 +0700 1022)         use_distinct = False
470a9bb22d4 (Loic Bistuer             2013-08-04 17:18:17 +0700 1023)         search_fields = self.get_search_fields(request)
470a9bb22d4 (Loic Bistuer             2013-08-04 17:18:17 +0700 1024)         if search_fields and search_term:
470a9bb22d4 (Loic Bistuer             2013-08-04 17:18:17 +0700 1025)             orm_lookups = [construct_search(str(search_field))
470a9bb22d4 (Loic Bistuer             2013-08-04 17:18:17 +0700 1026)                            for search_field in search_fields]
26a413507ab (Alix                     2020-05-26 17:53:18 +0200 1027)             for bit in smart_split(search_term):
26a413507ab (Alix                     2020-05-26 17:53:18 +0200 1028)                 if bit.startswith(('"', "'")):
26a413507ab (Alix                     2020-05-26 17:53:18 +0200 1029)                     bit = unescape_string_literal(bit)
a8df8e34f9f (Simon Charette           2013-10-07 13:38:15 -0400 1030)                 or_queries = [models.Q(**{orm_lookup: bit})
a8df8e34f9f (Simon Charette           2013-10-07 13:38:15 -0400 1031)                               for orm_lookup in orm_lookups]
a8df8e34f9f (Simon Charette           2013-10-07 13:38:15 -0400 1032)                 queryset = queryset.filter(reduce(operator.or_, or_queries))
4c599ece57f (Дилян Палаузов           2017-12-27 00:14:12 +0530 1033)             use_distinct |= any(lookup_needs_distinct(self.opts, search_spec) for search_spec in orm_lookups)
470a9bb22d4 (Loic Bistuer             2013-08-04 17:18:17 +0700 1034)
470a9bb22d4 (Loic Bistuer             2013-08-04 17:18:17 +0700 1035)         return queryset, use_distinct
470a9bb22d4 (Loic Bistuer             2013-08-04 17:18:17 +0700 1036)
```

So maybe no need for unit testing the queryset magic because that's done in django, but worth checking if there's any update that's relevant to this patch.

Still not sure how this will play out with #1170, but it seems inheriting from the same class, it's just that apparently the view gets a declaration per-field in DAL 4.0 so either both v3 and v4 declarations will be supported either only v4, but I'd tend to go with the former and maintain BC.